godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/Rect2.kt

```diff
@@ -0,0 +1,263 @@
+        val g = Rect2(this.position, this.size)
+        g.position.x -= left
+        g.position.y -= top
+        g.size.x += left + right
+        g.size.y += top + bottom
+        return g
+    }
+
+    /**
+     * Returns a copy of the Rect2 grown a given amount of units towards all the sides.
+     */
+    fun growMargin (margin: Margin, by: RealT ): Rect2 {
+        val g = Rect2(this.position, this.size)
+        g.position.x -= by
+        g.position.y -= by
+        g.size.x += by * 2
+        g.size.y += by * 2
+        return g
+    }
+
+    /**
+     * Returns true if the Rect2 is flat or empty.
+     */
+    fun hasNoArea(): Boolean {
+        return size.x <= 0 || size.y <= 0
+    }
+
+    /**
+     * Returns true if the Rect2 contains a point.
+     */
+    fun hasPoint(point: Vector2): Boolean {
+        return when {
+            point.x < position.x                -> false
+            point.y < position.y                -> false
+            point.x >= (position.x + size.x)    -> false
+            point.y >= (position.y + size.y)    -> false
+            else                                -> true
+        }
+    }
+
+    /**
+     * Returns true if the Rect2 overlaps with b (i.e. they have at least one point in common).
+     * If include_borders is true, they will also be considered overlapping if their borders touch, even without intersection.
+     */
+    fun intersects(rect: Rect2): Boolean {
+        return when {
+            position.x >= (rect.position.x + rect.size.x)   -> false
+            (position.x + size.x) <= rect.position.x        -> false
+            position.y >= (rect.position.y + rect.size.y)   -> false
+            (position.y + size.y) <= rect.position.y        -> false
+            else                                            -> true
+        }
+    }
+
+    /**
+     * Returns true if this Rect2 and rect are approximately equal, by calling is_equal_approx on each component.
+     */
+    fun isEqualApprox(rect: Rect2): Boolean {
+        return rect.position.isEqualApprox(this.position) && rect.size.isEqualApprox(this.size)
+    }
+
+    /**
+     * Returns a larger Rect2 that contains this Rect2 and b.
+     */
+    fun merge(rect: Rect2): Rect2 {
+        val ret = Rect2()
+
+        ret.position.x = min(rect.position.x, position.x)
+        ret.position.y = min(rect.position.y, position.y)
+
+        ret.size.x = max(rect.position.x + rect.size.x, position.x + size.x)
+        ret.size.y = max(rect.position.y + rect.size.y, position.y + size.y)
+
+        ret.size = rect.size - rect.position //make relative again
+
+        return ret
+    }
+
+
+    //UTILITIES
+    override fun equals(other: Any?): Boolean {
+        return when (other) {
+            is Rect2 -> position == other.position && size == other.size
+            else -> false
+        }
+    }
+
+    override fun toString(): String {
+        return "$position, $size"
```

Comment:
Shouldn't we do something like `"Rect2[position = $position, size = $size]"` or something like that? otherwise it's hard to read the toString when put inside a log statement
Or is it like that in godot?

Reply:
eek![image](https://user-images.githubusercontent.com/31470327/86237358-e84e6c00-bb9b-11ea-95f9-e1ed6332d13d.png)
This is like this in Godot. The toString function never return the type of the object.
We can discuss that this weekend to decide if we stick to Godot or replace it with a more specific string.
In the latter case, I will prepare a PR to fix all the core types (I will do one regardless because I noticed several mistakes I made in the previous cores).

Reply:
:+1: